// @neurosity/notion/dist/esm/api/firebase/FirebaseApp.js
import firebase from "firebase/app";
import "firebase/database";
import "firebase/auth";
import "firebase/functions";
import "firebase/firestore";
import { config } from "./config";
export const SERVER_TIMESTAMP = firebase.database.ServerValue.TIMESTAMP;
/**
 * @hidden
 */
export class FirebaseApp {
    constructor(options) {
        this.app = this.getApp(options.deviceId);
        this.standalone = this.app.name === options.deviceId;
    }
    getApp(deviceId) {
        const moduleApps = firebase.apps;
        const browserApps = typeof window !== "undefined" &&
            "firebase" in window &&
            "apps" in window.firebase
            ? window.firebase.apps
            : [];
        const neurosityApp = [...moduleApps, ...browserApps].find((app) => app.name === "[DEFAULT]" &&
            app.options.databaseURL === config.databaseURL);
        if (neurosityApp) {
            return neurosityApp;
        }
        if (deviceId) {
            const notionAppName = deviceId;
            const notionApp = moduleApps.find((app) => app.name === notionAppName);
            return notionApp
                ? notionApp
                : firebase.initializeApp(config, notionAppName);
        }
        return firebase.initializeApp(config);
    }
    goOnline() {
        this.app.database().goOnline();
    }
    goOffline() {
        this.app.database().goOffline();
    }
    disconnect() {
        if (this.standalone) {
            return this.app.delete();
        }
        return Promise.resolve();
    }
}
