// ipc-event-emitter/target/src/ipc-event-emitter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.IPCEventEmitter = undefined;

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

exports.default = IPC;

require('source-map-support/register');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _boolifyString = require('boolify-string');

var _boolifyString2 = _interopRequireDefault(_boolifyString);

var _events = require('events');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEBUG = (0, _boolifyString2.default)(process.env.IPC_EVENT_EMITTER_DEBUG);
var NODE_GE_4 = _semver2.default.gte(process.version, '4.0.0');
var OPTIONS = { debug: DEBUG };
var SEND_OR_DELIVER = NODE_GE_4 ? 'send' : 'deliver';
var TYPE = 'ipc-event-emitter';

function isIPCProcess($process) {
    return _lodash2.default.isFunction($process.send) && _lodash2.default.isFunction($process.on);
}

var IPCEventEmitter = exports.IPCEventEmitter = function (_EventEmitter) {
    (0, _inherits3.default)(IPCEventEmitter, _EventEmitter);

    function IPCEventEmitter($process) {
        var $options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
        (0, _classCallCheck3.default)(this, IPCEventEmitter);

        var _this = (0, _possibleConstructorReturn3.default)(this, (IPCEventEmitter.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter)).call(this));

        _this._pinned = {};

        if (isIPCProcess($process)) {
            _this.process = $process;
        } else {
            throw new TypeError('Invalid process; expected Process|ChildProcess, got: ' + $process);
        }

        var timeout = $options.timeout;

        if (timeout) {
            if (_lodash2.default.isNumber(timeout) && timeout > 0) {
                _this._timeout = timeout;
            } else {
                throw new TypeError('Invalid timeout; expected number > 0, got: ' + timeout);
            }
        }

        // register the listener for our IPC messages.
        // we distinguish our messages by the presence
        // of a `type` field with a value of TYPE (defined
        // above) e.g. { type: "ipc-event-emitter" }.
        // our messages come in the following flavours:
        //
        // unpin e.g. { unpin: 'ready' }
        //
        //     remove the named event from the target's
        //     `pinned` event object.
        //
        // pin e.g. { pin: true, emit: [ 'ready' ] }
        //
        //     add the event and its arguments to the target's
        //     `pinned` event object and fire the event in
        //     the target
        //
        // emit e.g. { emit: [ 'downloaded', 'http://example.com/file.txt' ] }
        //
        //     emit the specified event (e.g. `downloaded`) in the target
        $process.on('message', function (data) {
            if (data && data.type === TYPE) {
                if (data.unpin) {
                    delete _this._pinned[data.unpin];
                } else {
                    // XXX make sure we pin before we emit
                    // in case e.g. listener 1 registers
                    // listener 2 for this event while listener
                    // 1 is running (TODO test this).
                    if (data.pin) {
                        var _data$emit = (0, _toArray3.default)(data.emit),
                            name = _data$emit[0],
                            args = _data$emit.slice(1);

                        _this._pinned[name] = args;
                    }

                    (0, _get3.default)(IPCEventEmitter.prototype.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter.prototype), 'emit', _this).apply(_this, data.emit);
                }
            }
        });

        var options = _lodash2.default.assign({}, OPTIONS, $options);

        if (options.debug) {
            var EmitLogger = require('emit-logger');
            var name = options.name || $process.pid;
            var logger = new EmitLogger();

            logger.add($process, { name: name });
        }
        return _this;
    }

    // emit an event in the same way as `EventEmitter.emit`,
    // but emit it in the process at the other end of the IPC channel


    (0, _createClass3.default)(IPCEventEmitter, [{
        key: 'emit',
        value: function emit() {
            for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
                args[_key] = arguments[_key];
            }

            return this._sendAsync({ emit: args });
        }

        // register a sticky event
        //
        // a) emit the event in the IPC wrapper in the
        // target process (i.e. the child/parent process
        // at the other end of the IPC channel)
        //
        // b) the message handler in the target process
        // adds the event name (string) => arguments (array)
        // pair to its local map of pinned events

    }, {
        key: 'pin',
        value: function pin(name) {
            if (name === 'error') {
                throw new TypeError('"error" events cannot be pinned');
            }

            for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
                args[_key2 - 1] = arguments[_key2];
            }

            return this._sendAsync({ emit: [name].concat(args), pin: true });
        }

        // unregister a sticky event

    }, {
        key: 'unpin',
        value: function unpin(name) {
            return this._sendAsync({ unpin: name });
        }

        // append a new listener for this event

    }, {
        key: 'addListener',
        value: function addListener(name, listener) {
            (0, _get3.default)(IPCEventEmitter.prototype.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter.prototype), 'addListener', this).call(this, name, listener);
            return this._invokeIfPinned(name, listener);
        }

        // prepend a new listener for this event

    }, {
        key: 'prependListener',
        value: function prependListener(name, listener) {
            (0, _get3.default)(IPCEventEmitter.prototype.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter.prototype), 'prependListener', this).call(this, name, listener);
            return this._invokeIfPinned(name, listener);
        }

        // append a new listener for this event.
        // in theory, this should behave the same as
        // `addListener`. in practice, the two methods
        // may (for some reason) be implemented separately,
        // so rather than assuming or hoping, we just
        // delegate to the specified method

    }, {
        key: 'on',
        value: function on(name, listener) {
            (0, _get3.default)(IPCEventEmitter.prototype.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter.prototype), 'on', this).call(this, name, listener);
            return this._invokeIfPinned(name, listener);
        }

        // append a new one-time listener for this event

    }, {
        key: 'once',
        value: function once(name, listener) {
            return this._invokeIfPinned(name, listener, 'once');
        }

        // prepend a new one-time listener for this event

    }, {
        key: 'prependOnceListener',
        value: function prependOnceListener(name, listener) {
            return this._invokeIfPinned(name, listener, 'prependOnceListener');
        }

        // immediately fire a listener registered
        // with `on`, `addListener` &c. if its event
        // has been pinned.
        //
        // in most cases, the listener will already have
        // been registered before this method is called,
        // but for one-shot listeners (`once` and
        // `prependOnceListener`) we either a) fire them once
        // and forget them if the event is pinned (i.e. no
        // need to register them) or b) register them as usual
        // with super.once(...) or super.prependOnceListener(...)
        // if the event isn't pinned.
        //
        // to support the latter case, the name of the method
        // to delegate to is passed as the third parameter.

    }, {
        key: '_invokeIfPinned',
        value: function _invokeIfPinned(name, listener, delegate) {
            var pinned = this._pinned[name];

            if (pinned) {
                listener.apply(this, pinned);
            } else if (delegate) {
                // e.g. super.once(name, listener)
                (0, _get3.default)(IPCEventEmitter.prototype.__proto__ || (0, _getPrototypeOf2.default)(IPCEventEmitter.prototype), delegate, this).call(this, name, listener);
            }

            return this;
        }

        // use Node's underlying IPC protocol to send a
        // message between a parent process and a child
        // process
        //
        // # child.js
        // process.send(message) // child -> parent
        //
        // # parent.js
        // child = child_process.fork("child.js")
        // child.send(message)   // parent -> child

    }, {
        key: '_sendAsync',
        value: function _sendAsync(message) {
            var _this2 = this;

            var promise = void 0;

            // it's safe to mutate this message object as it's
            // always constructed internally
            message.type = TYPE;

            if (NODE_GE_4) {
                promise = _bluebird2.default.fromNode(function (callback) {
                    _this2.process.send(message, callback);
                });
            } else {
                promise = new _bluebird2.default(function (resolve, reject) {
                    process.nextTick(function () {
                        try {
                            resolve(_this2.process.send(message));
                        } catch (e) {
                            reject(e);
                        }
                    });
                });
            }

            var timeout = this._timeout;

            if (timeout) {
                return promise.timeout(timeout, 'IPC message took > ' + timeout + ' ms to ' + SEND_OR_DELIVER);
            } else {
                return promise;
            }
        }
    }]);
    return IPCEventEmitter;
}(_events.EventEmitter);

function IPC($process) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    return new IPCEventEmitter($process, options);
}
