// @neurosity/notion/dist/cjs/api/websocket/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const socket_io_client_1 = __importDefault(require("socket.io-client"));
const defaultOptions = {
    secure: true
};
/**
 * @hidden
 */
class WebsocketClient {
    constructor(options) {
        this.options = Object.freeze(Object.assign(Object.assign({}, defaultOptions), options));
        this.socket = socket_io_client_1.default(this.options.socketUrl, {
            path: `/${this.options.deviceId}`
        });
    }
    onMetric(subscription, callback) {
        return this.socket.on(`metrics/${subscription.id}`, callback);
    }
    removeMetricListener(subscription, listener) {
        this.socket.off(`metrics/${subscription.id}`, listener);
    }
    disconnect() {
        this.socket.removeAllListeners();
        if (this.socket.connected) {
            this.socket.disconnect();
        }
    }
}
exports.WebsocketClient = WebsocketClient;
WebsocketClient.serverType = "websocket";
